sim-administration/es2plus4dropwizard/src/main/kotlin/org/ostelco/sim/es2plus/Es2PlusClient.kt

```diff
@@ -30,90 +30,116 @@ class ES2PlusClient(
 
-        /// XXX TODO:
-        //       We  currently need jersey client for integration test and  httpClient for functional
-        //       SSL.  This is unfortunate, but also seems to be the shortest path towards a
-        //       functioning & testable  ES2+ client.   Should be implemented using two different
-        //       methods, that should then share a lot of common code.
+        val returnedContentType = response.getFirstHeader("Content-Type")
+        val expectedContentType = MediaType.APPLICATION_JSON
 
-        if (httpClient != null) {
+        if (returnedContentType.value != expectedContentType) {
+            throw ES2PlusClientException("Expected header Content-Type to be '$expectedContentType' but was '$returnedContentType'")
+        }
+        return ObjectMapper().readValue(response.entity.content, returnValueClass)
+                ?: throw ES2PlusClientException("null return value")
+    }
 
-            val objectMapper = ObjectMapper()
-            val payload = objectMapper.writeValueAsString(es2ProtocolPayload)
+    private fun <T> getEs2PlusHttpPostResponse(path: String, es2ProtocolPayload: T, expectedStatusCode: Int): HttpResponse {
+        if (httpClient == null) {
+            throw ES2PlusClientException("Attempt to use http client, even though it is not present in the client.")
+        }
 
-            val url = url(path)
-            val req = HttpPost(url)
+        val url = constructUrl(path)
+        val req = HttpPost(url)
 
-            req.setHeader("User-Agent", "gsma-rsp-lpad")
-            req.setHeader("X-Admin-Protocol", X_ADMIN_PROTOCOL_HEADER_VALUE)
-            req.setHeader("Content-Type", MediaType.APPLICATION_JSON)
-            req.setHeader("Accept", "application/json")
-            req.setHeader("Content-type", "application/json")
-            req.entity = StringEntity(payload)
+        val objectMapper = ObjectMapper()
+        val payload = objectMapper.writeValueAsString(es2ProtocolPayload)
 
-            val result: HttpResponse = httpClient.execute(req)
-                    ?: throw ES2PlusClientException("Null response from http httpClient")
+        req.setHeader("User-Agent", CLIENT_USER_AGENT)
+        req.setHeader("X-Admin-Protocol", X_ADMIN_PROTOCOL_HEADER_VALUE)
+        req.setHeader("Content-Type", MediaType.APPLICATION_JSON)
+        req.setHeader("Accept", MediaType.APPLICATION_JSON)
+        req.entity = StringEntity(payload)
 
-            // Validate returned response
-            val statusCode = result.statusLine.statusCode
-            if (expectedStatusCode != statusCode) {
-                val msg = "Expected return value $expectedStatusCode, but got $statusCode.  Body was \"${result.entity.content}\""
-                throw ES2PlusClientException(msg)
-            }
+        val response: HttpResponse = httpClient.execute(req)
+                ?: throw ES2PlusClientException("Null response from http httpClient")
```

Comment:
A few `throw`statements in the code. Should be rewritten to use `Either` I think. :-)

Reply:
This library does not use either.  A wrapper that does use either has been written in the ProfileVendorAdapter, it may be moved into the client once it is a bit more mature (it's still not compact enough).